### PR TITLE
refactor: extract semantic graph module into backend package (#303)

### PR DESCRIPTION
**Pull request by ibenian**
## Summary

Extracts the semantic graph parsing pipeline from `server.py` and `scripts/latex_to_graph.py` into a clean `backend/semantic_graph/` Python package, completing all 15 steps of the design doc.

- **New package**: `backend/semantic_graph/` with 8 focused modules — `constants.py`, `preprocessor.py`, `postprocessor.py`, `sympy_translator.py`, `equation_chain.py`, `cache.py`, `service.py`, `preprocess_result.py`
- **Single entry point**: `SemanticGraphService.derive(latex, domain)` orchestrates the full pipeline
- **server.py slimmed**: ~3,200 lines of inline parsing logic replaced with `SemanticGraphService` import; `_load_script_module` dynamic loader removed entirely
- **scripts/latex_to_graph.py**: reduced to a thin CLI wrapper delegating to the service
- **Test migration**: all semantic graph tests moved to `tests/backend/semantic_graph/`, 10 duplicate tests removed, 616 tests passing
- **Re-export shims removed**: old `models/` and `agents/` top-level shims deleted, all imports updated to `backend.*`
- **Design doc**: comprehensive test strategy added — three-layer architecture (unit/integration/domain suites), 28 domain-specific test suites covering math and physics areas, and test data generation techniques (Hypothesis, mutation testing, fuzzing, snapshots, boundary value analysis)

## Changes

| Area | What changed |
|------|-------------|
| `backend/semantic_graph/` | New 8-module package extracted from server.py and latex_to_graph.py |
| `backend/model/`, `backend/agents/` | Moved from top-level `models/`, `agents/` |
| `server.py` | Replaced ~3,200 lines with service import; removed `_load_script_module` |
| `scripts/latex_to_graph.py` | Slimmed to CLI wrapper |
| `scripts/__init__.py` | Created to enable `from scripts import ...` |
| `tests/backend/semantic_graph/` | 8 test files migrated + deduplicated |
| `docs/semantic-graph-module-design.md` | Design doc with architecture, migration plan, and test strategy |

## Test plan

- [x] All 616 tests pass (`./run.sh -m pytest`)
- [x] No remaining references to removed `_load_script_module`
- [x] No duplicate tests between old and new test locations
- [x] `from backend.semantic_graph.service import SemanticGraphService` works
- [ ] Reviewer: verify `server.py` endpoints still work end-to-end

Closes #303

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

### CODE REVIEW

File: backend/semantic_graph/sympy_translator.py [hunk 1]
```diff
@@ -0,0 +1,1404 @@
...
+        content = m.group(1).strip()
+        if content not in seen:
+            idx = len(seen)
+            seen[content] = idx
+            overrides[f"Xi_{{{idx}}}"] = {
+                "label": content,
+                "latex": r"\text{" + content + "}",
+                "type": "annotation",
+            }
+        return rf"\Xi_{seen[content]}"
```
[the remark this answers is not in the document]
Comment (ibenian) [reply backend/semantic_graph/sympy_translator.py]: Fixed in 9f41bb4 — placeholders now use fully braced form (`\\Xi_{N}`) and the node type is set to `"text"` instead of `"annotation"`. Tests updated to match.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

File: backend/semantic_graph/sympy_translator.py [hunk 2]
```diff
@@ -0,0 +1,1404 @@
...
+
+    latex = re.sub(
+        r"\\frac\{(d|\\partial)\^(?:\{(\d+)\}|(\d+))([^{}\d][^{}]*)\}\{\1\s*([^{}\s]+)\s*\^(?:\{\d+\}|\d+)\}",
+        _expand_higher_deriv,
+        latex,
+    )
+    latex = re.sub(r"\\ddot\{([^{}]+)\}", r"\\frac{d}{dt}\\frac{d \1}{d t}", latex)
+    latex = re.sub(r"\\dot\{([^{}]+)\}", r"\\frac{d \1}{d t}", latex)
+    latex = re.sub(r"_([A-Za-z0-9])(?![A-Za-z0-9_{])", r"_{\1}", latex)
+    latex = re.sub(r"\\(?:quad|qquad|,|;|!)\s*", " ", latex)
```
[the remark this answers is not in the document]
Comment (ibenian) [reply backend/semantic_graph/sympy_translator.py]: Fixed in 9f41bb4 — the hand-rolled dot/ddot regexes are replaced with `LaTeXPreprocessor.rewrite_dot_derivatives()` + `normalize_frac_derivatives()`, which handle `\\dddot`/`\\ddddot`, trailing subscripts/superscripts, and emit the canonical `\\frac{d}{d t}` form that SymPy reliably parses.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

File: scripts/render_math.py [hunk 3]
```diff
@@ -32,11 +32,11 @@
...
 from typing import Any
 
+from backend.semantic_graph.sympy_translator import latex_to_semantic_graph
+
 try:
-    from latex_to_graph import latex_to_semantic_graph
     from graph_to_mermaid import semantic_graph_to_mermaid, load_theme, validate_graph
 except ImportError:
-    from scripts.latex_to_graph import latex_to_semantic_graph
     from scripts.graph_to_mermaid import semantic_graph_to_mermaid, load_theme, validate_graph
```
[the remark this answers is not in the document]
Comment (ibenian) [reply scripts/render_math.py:41]: Fixed in 9f41bb4 — `run.sh` now exports `PYTHONPATH` to the repo root before exec, so `backend.*` imports resolve correctly when any script is run by path.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

File: scripts/latex_to_graph.py [hunk 4]
```diff
@@ -29,2196 +28,43 @@
...
-    (r"\lt", {"op": "less_than", "label": "less than", "emoji": "<"}),
-]
-
-_STYLE_SYMBOL_COMMAND_RE = re.compile(
-    r"\\(?P<style>mathbb|mathbf|mathcal|mathfrak|mathscr|mathrm)\s*"
-    r"\{(?P<body>[^{}]+)\}"
+from backend.semantic_graph.sympy_translator import (  # noqa: F401
+    latex_to_semantic_graph,
+    parse_var_overrides,
+    operator_kind,
```
[the remark this answers is not in the document]
Comment (ibenian) [reply scripts/latex_to_graph.py:40]: Fixed in 9f41bb4 — same `run.sh` PYTHONPATH fix covers this script too. The repo root is always on `PYTHONPATH` now, so `backend.*` imports work when run via `./run.sh scripts/latex_to_graph.py`.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

File: backend/semantic_graph/preprocessor.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+        spacing = r"(?:\s|\\quad|\\qquad|\\,|\\;|\\!|\\:)*"
+        pattern = re.compile(
+            spacing + r"\(([^()]*\\text\{[^}]+\}[^()]*)\)\s*$"
+        )
+        while True:
+            m = pattern.search(latex)
+            if not m:
+                break
+            inner = m.group(1).strip()
+            label = re.sub(r"\\text\{([^}]+)\}", r"\1", inner)
```
[the remark this answers is not in the document]
Comment (ibenian) [reply backend/semantic_graph/preprocessor.py]: Resolved in commits 354bc75 → ccd2697 → 3bb00d0. All polynomial regex patterns were replaced with non-backtracking alternatives or possessive quantifiers. Zero open CodeQL alerts remain on this PR.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

File: backend/semantic_graph/sympy_translator.py [hunk 6]
```diff
@@ -0,0 +1,1371 @@
...
+            idx = len(seen)
+            seen[content] = idx
+            overrides[f"Xi_{{{idx}}}"] = {
+                "label": content,
+                "latex": r"\text{" + content + "}",
+                "type": "annotation",
+            }
+        return rf"\Xi_{seen[content]}"
+
+    rewritten = re.sub(r"\\text\{([^}]+)\}", repl, latex)
```
[the remark this answers is not in the document]
Comment (ibenian) [reply backend/semantic_graph/sympy_translator.py]: Resolved — same regex fix series (354bc75 → 3bb00d0). The `\\text{}` matching pattern was rewritten to avoid polynomial backtracking.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

File: backend/semantic_graph/sympy_translator.py [hunk 7]
```diff
@@ -0,0 +1,1371 @@
...
+        return rf"\Phi_{{{seen[full]}}}"
+
+    braket_pat = (
+        r"(?:\\left\s*)?\\langle\s*"
+        r"([^|]*?)"
+        r"\s*\|\s*"
+        r"([^|]*?)"
+        r"\s*\\rangle(?:\s*\\right\s*\.)?"
+    )
+    rewritten = re.sub(braket_pat, _repl_braket, latex)
```
[the remark this answers is not in the document]
Comment (ibenian) [reply backend/semantic_graph/sympy_translator.py]: Resolved in the ReDoS fix series. The braket/langle pattern was rewritten to eliminate polynomial backtracking on whitespace repetitions.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

File: backend/semantic_graph/sympy_translator.py [hunk 8]
```diff
@@ -0,0 +1,1371 @@
...
+    spacing = r"(?:\s|\\quad|\\qquad|\\,|\\;|\\!|\\:)*"
+    pattern = re.compile(
+        spacing + r"\(([^()]*\\text\{[^}]+\}[^()]*)\)\s*$"
+    )
+    while True:
+        m = pattern.search(latex)
+        if not m:
+            break
+        inner = m.group(1).strip()
+        label = re.sub(r"\\text\{([^}]+)\}", r"\1", inner)
```
[the remark this answers is not in the document]
Comment (ibenian) [reply backend/semantic_graph/sympy_translator.py]: Resolved — covered by the same `\\text{}` regex rewrite in the fix series.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

File: backend/semantic_graph/sympy_translator.py [hunk 9]
```diff
@@ -0,0 +1,1371 @@
...
+        if order <= 1:
+            return m.group(0)
+        wrapper = r"\frac{%s}{%s %s}" % (op, op, var)
+        core = r"\frac{%s %s}{%s %s}" % (op, func, op, var)
+        return wrapper * (order - 1) + core
+
+    latex = re.sub(
+        r"\\frac\{(d|\\partial)\^(?:\{(\d+)\}|(\d+))\s*([^}]+)\}\{\1\s*([^}]+?)\s*\^(?:\{\d+\}|\d+)\}",
+        _expand_higher_deriv,
+        latex,
```
[the remark this answers is not in the document]
Comment (ibenian) [reply backend/semantic_graph/sympy_translator.py:388]: Resolved — the higher-order derivative `\\frac{d^N ...}` pattern was rewritten to avoid polynomial backtracking on whitespace.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

File: backend/semantic_graph/sympy_translator.py [hunk 10]
```diff
@@ -0,0 +1,1371 @@
...
+    )
+    latex = re.sub(r"\\ddot\{([^}]+)\}", r"\\frac{d}{dt}\\frac{d \1}{d t}", latex)
```
[the remark this answers is not in the document]
Comment (ibenian) [reply backend/semantic_graph/sympy_translator.py]: Resolved — the `\\ddot{}` regex was replaced by `LaTeXPreprocessor.rewrite_dot_derivatives()` in commit 9f41bb4 (Copilot review fix), which also eliminates this ReDoS vector.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

File: backend/semantic_graph/sympy_translator.py [hunk 11]
```diff
@@ -0,0 +1,1371 @@
...
+    latex = re.sub(r"\\dot\{([^}]+)\}", r"\\frac{d \1}{d t}", latex)
```
[the remark this answers is not in the document]
Comment (ibenian) [reply backend/semantic_graph/sympy_translator.py]: Resolved — same as the `\\ddot` fix above. Both dot-notation regexes replaced by `LaTeXPreprocessor` methods.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>

File: backend/semantic_graph/sympy_translator.py [hunk 12]
```diff
@@ -0,0 +1,1371 @@
...
+        if order <= 1:
+            return m.group(0)
+        wrapper = r"\frac{%s}{%s %s}" % (op, op, var)
+        core = r"\frac{%s %s}{%s %s}" % (op, func, op, var)
+        return wrapper * (order - 1) + core
+
+    latex = re.sub(
+        r"\\frac\{(d|\\partial)\^(?:\{(\d+)\}|(\d+))\s*([^{}]+)\}\{\1\s*([^{}\s]+)\s*\^(?:\{\d+\}|\d+)\}",
+        _expand_higher_deriv,
+        latex,
```
[the remark this answers is not in the document]
Comment (ibenian) [reply backend/semantic_graph/sympy_translator.py:388]: Duplicate of #5 above — resolved in the same ReDoS fix series.

🤖 Co-Authored-By: Claude <81847+claude@users.noreply.github.com>
